### DISCUSSION TITLE
Difference between PV string power sum and "Inverter Input power" -> Efficency loss?

### DISCUSSION THREAD

**Discussion by SvenAbels** [wlcrs/huawei_solar#540] [General]
Hi,

I got a sun2000 inverter. When multiplying the voltage and the current of both strings and creating the sum of them, I should get the total power (Watt) amount. At least this is my understanding.   -   But I noticed that this sum is always above the "Inverter Input power" value from the inverter. 

The difference between the values and the "Inverter Input power" is fluctuating but it get's near 100% when a lot of PV power is consumed and moves towards 90% on low production times.   I wonder if this difference is due to the inverter efficency and indeed when looking at the sun2000 data sheet the values would roughly fit.

However, if this would be the case then I wonder why the "sensor.input_power_with_efficiency_loss" should be calculated when calculating the total energy production as described in the FAQs. Because of the efficency is already inside the "Inverter Input power"  values then this whould not be necessary...

Best regards,

Sven

**Comment by SvenAbels** [comment]
@wlcrs Do you have an idea about this?

**Reply by wlcrs** [reply]
No, sorry

**Comment by SvenAbels** [comment]
An update on this: I did some more tests to analyze this and my current colclusion is that the Rieman sum of the sensor.inverter_input_power is indeed already considering the efficency loss, so for calculating the solar power yield today, it is not needed to calculate the efficency loss.

The following diagram is showing my measurements from today. It was a very cloudy morning and no battery was charged and I only got a small amount of power generated so there should be a lot of efficency loss. 

![image](https://github.com/wlcrs/huawei_solar/assets/1993969/f14f0a47-98fb-4afa-801a-9cbe126ebc30)


The purple line shows the theoretical watt amount by multiplyin Volt and Ampere of the 2 strings.  The light-blue line shows the measurement of an external power meter which I own from my wallbox and which measures the watt amount coming from the inverter to my house.  

The dark blue line shows the plain sensor.inverter_input_power and the orange line shows the inverter input power with the efficency impace as shown in the WLCRS FAQ.   As you can see, the dark blue line (i.e. the original  sensor.inverter_input_power) is almost identical with the reality while I assume that the difference between the purple and the dark blue line is the real efficency (which is very big due to the small amoutn of Watt getting into my 10 kWp inverter).

**Comment by JensenNick** [comment]
@SvenAbels My understanding of the sensor.input_power_with_efficiency_loss is that it is supposed to emulate the yield you can see in the Fusion Solar app. The yield shown in the app is being corrected by the loss you have in your charging and discharging of your battery.

@wlcrs Could you please have a look at the wiki I have added - mostly just not trying to mess things up for you / contribute.

**Reply by wlcrs** [reply]
I've removed the separate page, and created a section "Related projects" on the home page of the wiki instead. I've tried to describe your project in one paragraph, feel free to tweak it a bit. I feel that this way, people quickly get the gist of what your work is about. A separate page made it harder to find. 

All the warnings, disclaimers and doubts should not be included here: you are trying to make people enthousiastic to read your work, not scare them away. What you did and how you documented is nicely done. You should be proud of your work, and not downplay it before people have even seen it!

**Reply by SvenAbels** [reply]
@JensenNick From what I can see, the Fusion Solar yield display matches the last method of the https://github.com/wlcrs/huawei_solar/wiki/Daily-Solar-Yield page, i.e. the calculation witout the input_power_with_efficiency_loss.   

As far as I understand it, the input_power_with_efficiency_loss tries to estimate the efficiency loss of the inverter and not the battery charging because it is only applied to the sensor.inverter_input_power and does not take into account any battery value charging/discharging.

**Reply by JensenNick** [reply]
@wlcrs Thank you and thanks for you initial view on my project. Warnings and disclaimers :-), just didn't want people to expect too much and to realize were I'm coming from. But cool, lets see how it goes.

**Reply by JensenNick** [reply]
@SvenAbels That may very well be the case. I kind of got caught up in my own project and only had focus on what i felt I needed and wanted from what was provided at the time - so I may very well be remembering wrong or the output could have been changed.

If I may what is your goal? I ask because I have made some sensors in the project mentioned above, and in addition to these I have one tracking the efficiency of the battery. I believe this is where you'll se the greatest loss. My numbers indicate that the efficiency of the battery is around 80 %.

**Reply by SvenAbels** [reply]
@JensenNick I only try to optimize my own measurements in home assistant to track the numbers as close as I can.  

Testing the battery efficency is also an important element. I also did some tests about the battery loss but I didn't get very deep into it so far. My tests indicate an efficency that is fluctuating between 77% and 90%. I don't know why it is changing so much but I assume that the efficency of the battery is connected to the SOC level of the battery.

**Reply by JensenNick** [reply]
@SvenAbels I have my own "project" which I have now shared in public. Take a look if you wish. The loss is highly dependent on the temperature in which the battery is placed (mine is in a 18 degree garage). Next is that according to my finding the charge process has a effociency of aprox. 87% and the discharge aprox. 93% - giving a total of aprox. 80%. On top of this my estimation is that there is a aprox. 5% loss pr. day (I monitored the loss for 19 hours which resulted in 0,4 kWh loss - 10 kWh battery charged to 87%). I'm also seeing an "after charge" for up to 6 hours after the charging has ended - I didn't see this after the latest firmware update though. The “total efficiency” is aprox. 75% for me.

Link to my projects
https://github.com/JensenNick?tab=repositories

Here is the sensor I use to track efficiency. I was goint to make aditional sensors, to see if the loss is greater close to 0% and 100%, but the charging and soc shows a very straight line. Probably a buffer in each end. I'm considering using 100% of the battery.

```
- trigger:
    - platform: template
      value_template: >
        {% set start_value_soc = 50 %}
          {{ states('sensor.battery_state_of_capacity_1') | float(0) == start_value_soc }}
  sensor:
    - name: "Battery Efficiency"
      unique_id: battery_efficiency
      unit_of_measurement: "%"
      state: >
        {% set start_value_discharge = 1.995 %}
        {% set start_value_charge = 5.032 %}
        {% set discharge = states('sensor.energy_battery_discharge') | float(0) - start_value_discharge %}
        {% set charge = states('sensor.energy_battery_charge') | float(1) - start_value_charge %}
          {{ (( discharge / charge ) * 100 ) | round(1) }}
```
[
![9AA40EE7-2E9B-4D8B-B688-8A4E043B92F9](https://github.com/wlcrs/huawei_solar/assets/149193835/38eee7ce-b56f-49b6-9ca1-5dd278655753)
](url)

**Reply by SvenAbels** [reply]
Hi,

this is really great work. :-)   Thanks a lot for sharing it via GitHub. 

The battery loss esm to fit well with my numbers, too but I need to do some more measurements to track it over time. 



Best regards,

Sven

**Reply by JensenNick** [reply]
@SvenAbels I've noticed a sudden increase in my efficiency (from 77% to approx.. 79%). I suspect it is because my input is more than 30 days old and therefore deleted in the history. I don't have a solution for this here and now and I'm kind of debating with my self what would be "correct" or should we have more sensors - a short term and a long term. I was thinking about a sensor for efficiency for grid charge and one for yield, but that is kind of impossible, since you can't tell much of the discharge comes from what. Any inputs are welcome. 

PS: Thinking this through again the result must be correct since it’s calculating on basis of the latest accumulated value. Am I wrong?
